lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,219 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.slice.SizeOf.estimatedSizeOf;
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.GIGABYTE;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_FILE_LOCATION;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_FILE_READ_SIZE;
+import static io.trino.filesystem.tracing.CacheSystemAttributes.CACHE_KEY;
+import static io.trino.filesystem.tracing.Tracing.withTracing;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public final class MemoryFileSystemCache
+        implements TrinoFileSystemCache
+{
+    private final Tracer tracer;
+    private final int maxContentLengthBytes;
+    private final Cache<String, Optional<Slice>> cache;
+    private final AtomicLong largeFileSkippedCount = new AtomicLong();
+
+    @Inject
+    public MemoryFileSystemCache(Tracer tracer, MemoryFileSystemCacheConfig config)
+    {
+        this(tracer, config.getCacheTTL(), config.getMaxSize(), config.getMaxContentLength());
+    }
+
+    public MemoryFileSystemCache(Tracer tracer, Duration expireAfterWrite, DataSize maxSize, DataSize maxContentLength)
+    {
+        this.tracer = requireNonNull(tracer, "tracer is null");
+        checkArgument(maxContentLength.compareTo(DataSize.of(1, GIGABYTE)) <= 0, "maxContentLength must be less than or equal to 1GB");
+        this.cache = EvictableCacheBuilder.newBuilder()
+                .maximumWeight(maxSize.toBytes())
+                .weigher((Weigher<String, Optional<Slice>>) (key, value) -> toIntExact(estimatedSizeOf(key) + sizeOf(value, Slice::getRetainedSize)))
+                .expireAfterWrite(expireAfterWrite.toMillis(), TimeUnit.MILLISECONDS)
+                .shareNothingWhenDisabled()
+                .recordStats()
+                .build();
+        this.maxContentLengthBytes = toIntExact(maxContentLength.toBytes());
+    }
+
+    @Override
+    public TrinoInput cacheInput(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateInput");
+            return withTracing(span, delegate::newInput);
+        }
+
+        return new MemoryInput(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
```

Comment:
getFromCache reads "get from cache if it is there"
in this context, incrementing something related to file size is weird, since file size wasn't checked

renaming getFromCache to "getOrLoadFromCache" could help a bit, because it would reveal loading
but it would not reveal that lack of loading implies file size was too big.

can we do even better?

Reply:
I could have moved incrementing the counter to `load` method, but then that would miss counting the number of times large file was requested from cache.
I've renamed method for now